### PR TITLE
Sign internals of EXE package so that it works correctly when signed

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Sign internals of EXE package so that it works correctly when signed
- Refactor exe package creation into separate function
- Add functions to enable signing of engine of the exe package
- Move signing of MSI to package creation Job

## PR Context

The extracted EXE and MSI of the wrapper must be signed if the EXE is signed or install will fail.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 1]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+        cd $(PowerShellRoot)
+        Import-Module $(PowerShellRoot)/build.psm1 -Force
+        Import-Module $(PowerShellRoot)/tools/packaging -Force
+
+        $msiPath = '$(Build.StagingDirectory)\signedPackages\PowerShell-$(version)-win-${{ parameters.architecture }}.msi'
+
+        New-ExePackage -ProductVersion '$(version)' -MsiLocationPath $msiPath -ProductTargetArchitecture ${{ parameters.architecture }}
+        $exePath = Get-ChildItem '.\PowerShell-*.exe' | Select-Object -First 1 -ExpandProperty fullname
+        $enginePath = Join-Path -Path '$(System.ArtifactsDirectory)\unsignedEngine' -ChildPath engine.exe
+        Dismount-ExePackageEngine -ExePath $exePath -EnginePath $enginePath
```
Comment (rjmholt) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:308]: I'm mostly not clear on the `Mount`/`Dismount-ExePackageEngine` uses here:

- Why is it `Mount`/`Dismount`? What does it leave on the filesystem?
- Why do we dismount in one task and then mount later in another?
- Are we depending on filesystem state between tasks?

**Review by rjmholt — COMMENTED** [review]
I'd like to get someone who understand this better to review it as well if I can. @adityapatwardhan, @anmenaga or possibly @andschwa.

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml (hunk 1 above)
Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:308]: It's more like an archive.... so perhaps... `Expand`/`Compress`

Comment (rjmholt) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:308]: That makes some sense. What is exe package engine? What does each command do? Can we add comments that contain the answers to those questions and ideally the questions above?

Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:308]: The engine of the exe is the installer enignine

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 2]
```diff
@@ -261,7 +263,81 @@ jobs:
         Write-Host "Uploading $packagePath"
         Write-Host "##vso[artifact.upload containerfolder=signed;artifactname=signed]$packagePath"
       }
-    displayName: Upload packages
+    displayName: Upload unsigned packages
+
+  - ${{ if and(ne(variables['BuildConfiguration'],'minSize'), startsWith(variables['Architecture'], 'x')) }}:
```
Comment (rjmholt) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:268]: This is probably more explicit

```suggestion
  - ${{ if and(ne(variables['BuildConfiguration'],'minSize'), in(variables['Architecture'], 'x64', 'x86')) }}:
```

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 3]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+          signOutputPath: $(System.ArtifactsDirectory)\signedEngine
+          certificateId: "CP-230012"
+          pattern: |
+            **\*.exe
+          useMinimatch: true
+          shouldSign: $(SHOULD_SIGN)
+          displayName: Sign Exe Wrapper Engine
+
+    - pwsh: |
+        cd $(PowerShellRoot)
```
Comment (rjmholt) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:323]: I'm guessing we need to make sure we handle whitespace in the path:
```suggestion
        cd '$(PowerShellRoot)'
```

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 4]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+        Import-Module $(PowerShellRoot)/tools/packaging -Force
+
+        $exePath = Get-ChildItem '.\PowerShell-*.exe' | Select-Object -First 1 -ExpandProperty fullname
+        $enginePath = Join-Path -Path '$(System.ArtifactsDirectory)\signedEngine' -ChildPath engine.exe
+        $enginePath | Get-AuthenticodeSignature | out-string | Write-Verbose -verbose
+        Mount-ExePackageEngine -ExePath $exePath -EnginePath $enginePath
+      displayName: Use signed engine in exe wrapper
+
+    - powershell: |
+        cd $(PowerShellRoot)
```
Comment (rjmholt) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:334]:
```suggestion
        cd '$(PowerShellRoot)'
```

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 5]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+          pattern: |
+            **\*.exe
+          useMinimatch: true
+          shouldSign: $(SHOULD_SIGN)
+          displayName: Sign Exe Wrapper Engine
+
+    - pwsh: |
+        cd $(PowerShellRoot)
+        Import-Module $(PowerShellRoot)/build.psm1 -Force
+        Import-Module $(PowerShellRoot)/tools/packaging -Force
```
Comment (rjmholt) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:325]:
```suggestion
        Import-Module '$(PowerShellRoot)/build.psm1' -Force
        Import-Module '$(PowerShellRoot)/tools/packaging' -Force
```

Comment (rjmholt) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:325]: Is there a reason to use `-Force` here? Are we changing the psm1s in the build process at all?

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml (hunk 1 above)
Comment (rjmholt) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:308]: Ok having read a bit about [insignia](https://wixtoolset.org/documentation/manual/v3/overview/insignia.html), here are my recommendations:

- Rename the functions to `Disconnect-`/`Connect-ExePackageEngine`
- Add a comment in the function definitions like `Needed to detach the WIX Burn engine from the exe for signing` and `Needed to reattach the WIX Burn engine to the exe after signing`
- Add a comment in the YAML-embedded script where we detach saying `Detach the packaging engine so we can sign the exe`
- Add a similar comment when we re-attach saying `Re-attach the packaging engine to the now-signed exe so we can create the MSI`

It's not clear from this file why we need to re-attach Burn at the end though. We seem to do the following:

- sign the exe
- re-attach Burn to the exe
- upload it (but the display name says *unsigned*??)

I don't follow this sequence.

**Comment by TravisEz13** [comment]
disconnect/connect are not data verbs

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 6]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+        Import-Module '$(PowerShellRoot)/build.psm1' -Force
+        Import-Module '$(PowerShellRoot)/tools/packaging' -Force
+
+        $exePath = Get-ChildItem '.\PowerShell-*.exe' | Select-Object -First 1 -ExpandProperty fullname
+        $enginePath = Join-Path -Path '$(System.ArtifactsDirectory)\signedEngine' -ChildPath engine.exe
+        $enginePath | Get-AuthenticodeSignature | out-string | Write-Verbose -verbose
+        Compress-ExePackageEngine -ExePath $exePath -EnginePath $enginePath
+      displayName: Use signed engine in exe wrapper
+
+    - powershell: |
```
Comment (adityapatwardhan) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:333]: any reason this is powershell not pwsh?

**Review by adityapatwardhan — APPROVED** [review]

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml [hunk 7]
```diff
@@ -261,7 +263,81 @@ jobs:
...
+          buildOutputPath: $(System.ArtifactsDirectory)\pkgSigned
+          signOutputPath: $(Build.StagingDirectory)\signedPackages
+          certificateId: "CP-230012"
+          pattern: |
+            **\*.msi
+          useMinimatch: true
+          shouldSign: $(SHOULD_SIGN)
+          displayName: Sign MSI
+
+    - powershell: |
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:280]:
```suggestion
    - pwsh: |
```

File: tools/releaseBuild/azureDevOps/templates/windows-packaging.yml (hunk 6 above)
Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:333]:
```suggestion
    - pwsh: |
```

Comment (TravisEz13) [reply tools/releaseBuild/azureDevOps/templates/windows-packaging.yml:333]: just hasn't been tested

**Pull request merged**